pkg/controller/resourceclaim/controller.go

```diff
@@ -161,80 +175,154 @@ func (ec *Controller) enqueuePod(logger klog.Logger, obj interface{}, deleted bo
 	pod, ok := obj.(*v1.Pod)
 	if !ok {
 		// Not a pod?!
+		logger.Error(nil, "enqueuePod called for unexpected object", "type", fmt.Sprintf("%T", obj))
 		return
 	}
 
-	if deleted {
-		ec.deletedObjects.Add(pod.UID)
-	}
-
 	if len(pod.Spec.ResourceClaims) == 0 {
 		// Nothing to do for it at all.
 		return
 	}
 
+	if deleted {
+		logger.V(5).Info("pod got deleted", "pod", klog.KObj(pod))
+		ec.deletedObjects.Add(pod.UID)
+	}
+
 	logger.V(6).Info("pod with resource claims changed", "pod", klog.KObj(pod), "deleted", deleted)
 
 	// Release reservations of a deleted or completed pod?
-	if deleted || isPodDone(pod) {
+	if reason := podDoesNotNeedClaims(pod, deleted); reason != "" {
 		for _, podClaim := range pod.Spec.ResourceClaims {
+			// In most cases, the pod probably held a reservation,
+			// so here we avoid looking up the claim to check. The
+			// worker will do that.
 			claimName := resourceclaim.Name(pod, &podClaim)
 			key := claimKeyPrefix + pod.Namespace + "/" + claimName
 			logger.V(6).Info("pod is deleted or done, process claim", "pod", klog.KObj(pod), "key", key)
 			ec.queue.Add(key)
 		}
 	}
 
-	// Create ResourceClaim for inline templates?
-	if pod.DeletionTimestamp == nil {
-		for _, podClaim := range pod.Spec.ResourceClaims {
+	needsWork, reason := ec.podNeedsWork(pod)
+	if needsWork {
+		logger.V(5).Info("enqueing pod", "pod", klog.KObj(pod), "reason", reason)
+		ec.queue.Add(podKeyPrefix + pod.Namespace + "/" + pod.Name)
+		return
+	}
+	logger.V(6).Info("not enqueing pod", "pod", klog.KObj(pod), "reason", reason)
+}
+
+func podDoesNotNeedClaims(pod *v1.Pod, deleted bool) string {
```

Comment:
nit: Does returning `bool, string` here make sense? 

At the callsite we would have 
```
if notNeeded, reason := podDoesNotNeedClaims(); notNeeded {

}
```
which may make the intent clearer than checking for an empty string.

Reply:
(I see you use this for `podNeedsWork` below)

Reply:
Yes, I should use the same pattern also here. I'll change it.